### DISCUSSION TITLE
A Production instance of M.I.L.E.S

### DISCUSSION THREAD

**Discussion by asleepynerd** [small-cactus/M.I.L.E.S#11] [Ideas]
I have a few powerful cloud servers, I was thinking, rather than making the user download the whole git repo, we make a standalone application, that accesses a cloud server for all the responses (for a small fee/subscription) so it can be used on lower end devices. As for the whisper models, you can use services like https://assembly.ai which uses a live websocket connection, and sending the audio data to that instead of running the whisper model locally. This would also make it far easier to implement the application on devices, possibly making a "bundle" with a raspberry pi and a small display with speakers, sort of like an Amazon Echo. This idea could generate an income from those who want the product, but none of the hassle of having a sufficient device, or needing to do the manual setup, albeit MacOS is easier to setup. And for people who have the resources/time for setting up the product, then they can use the open source version of M.I.L.E.S.

**Comment by small-cactus** [comment]
The only thing ran on device is the wake word model, and the whisper processing. The wake word model could run on literally anything, it is extremely efficient, as for whisper, that’s a different story.

A standalone app or hardware device that has a subscription to use would not get very many users. One of the best selling points for Alexa and the Google home is that they’re cheap. With the current cost rates of average usage for Miles, you’d have to bill the customer monthly a price of near $10. At that price the hardware would literally have to be free or near free.

One of the reasons I made Miles was to give a solution to terrible voice assistants like Siri, the idea was to have a standalone app or hardware, those ideas are almost impossible with my current implementation, the entire logic is written in Python, meaning it’s extremely hard to package the app as 1 exe file. And there’s no way to input data into the Python file while it’s running without importing tons of extra files.

As for pricing, Miles can only do 50% of what a normal voice assistant can do, and with a 20% longer delay. Sometimes it makes up for this in answering questions that other voice assistants can’t, but Miles lacks several must have features to be useful as a paid service.

But I’d still like to hear your take on this. All I know is that I’d never pay for this, it’s too useless, hence why it’s here on GitHub for free.

**Reply by asleepynerd** [reply]
On Packaging as a Binary:
While it's improbable that it's running a python server with an electron frontend, it's not impossible. You could possibly create a installer that installs to a folder, creates a shortcut and have it as an "application". However this only really applies to windows systems and I know that other people use both Linux and MacOS to run M.I.L.E.S. Another possible solution to implement all this is a singular install script. Having an install script means you can have 1 specific script for both MacOS and Linux based systems, and another specific to windows. Using a bash script allows you to execute commands on the users behalf, so this allows the script to check for the operating system and act accordingly. As for making it look like a standalone application however, is definitely difficult with this method. There's an app available for MacOS by the name "Web2Mac" (https://web2mac.app/) that allows users to install "web pages" and render them as an app in the launch pad. This perhaps could be possible by removing the use of electron and exposing a local server.

On Business Model:
I understand your concern about the cost and agree that a high subscription fee could deter potential users. However, there might be alternative ways to monetize the project without putting the burden on the user:

1. Freemium Model: Provide basic features for free, and charge for premium features. This could include faster response times, priority access to new features, or advanced customization options.
2. Hardware Sales: If the Raspberry Pi bundle idea is feasible, the hardware itself could be a source of income. While the software remains free and open-source, users who want a hassle-free, ready-to-use device could purchase the bundle.
3. Donations and Sponsorships: As an open-source project, you could accept donations from users who find value in M.I.L.E.S. This could be facilitated through platforms like Patreon or GitHub Sponsors. Additionally, corporate sponsorships could be explored.

On Additional Functionality:
Consider integrating more functionality or plugins to enhance M.I.L.E.S.'s capabilities. This could include support for additional languages, integration with other services or APIs, like translation and so on.

On Mobile App:
Creating a mobile app could be a great way to make M.I.L.E.S. more accessible and easier to configure. The app could provide a user-friendly interface for setting up and customizing M.I.L.E.S., and could even serve as a remote control for the system.

Feel free to let me know if any of these ideas are feasible.

**Reply by small-cactus** [reply]
The only thing from that list I see as feasible would be hardware bundle, like what Home Assistant is doing.

**Reply by asleepynerd** [reply]
Is it not possible to make 1 line installer scripts?

**Reply by small-cactus** [reply]
It's not possible with my skill level, I could probably do it, but I'd have to figure out how to upload it to something like home brew or another thing like that for windows.

**Reply by asleepynerd** [reply]
Uploading to brew.sh wouldn’t be the issue, its how to setup symlinks to make sure that it installs properly on the users device and runs the npm start script, however it should be possible by just symlinking a bash script that checks for the required tools. if its there, it runs `npm start`, if not it’ll return an error saying they need to run `miles install` and so on. It should be relatively the same process for windows since you can use chocolatey as an alternative to brew for windows.
